doc/developer/platform/formalism.md

```diff
@@ -194,3 +199,111 @@ A view introduces *constraints* on the capabilities and frontiers of its input a
+The `CREATE TABLE` command introduces a new source that is not automatically populated by an external source, and is instead populated by `UpdateAndDowngrade` commands.
+The Adapter layer may use a write-ahead log to durably coalesce multiple writes at the same timestamp, as the `UpdateAndDowngrade` command does not otherwise allow this.
+The `DROP TABLE` command drops both the read and write capabilities for the source.
+
+## Indexes
+
+Materialize supports a `CREATE INDEX` command that results in a dataflow that computes and maintains the contents of the index's subject, in indexed representation.
+The `CREATE INDEX` command can be applied to any read-only view whose definition avoids certain non-deterministic functions (e.g. `now()`, `rand()`, environment variables).
+
+Materialize's main feature is that for any indexable view, one receives identical results whether re-evaluating the view from scratch or reading it from a created index.
+
+The `CREATE INDEX` command results in a `MaintainView` command for the Compute layer.
+The inputs to that command are `GlobalId` identifiers naming Storage or Compute collections.
+The command is issued with read capabilities all advanced to a common frontier (often: the maximum frontier of read capabilities in the input).
+The command returns a read capability for the indexed collection, which the Adapter layer maintains and perhaps regularly downgrades, as is its wont.
+The dataflow remains active until a corresponding `DROP INDEX` command is received by the Adapter layer, which discards its read capability for the indexed collection.
+
+## Transactions
+
+The Adapter layer provides interactive transactions through the `BEGIN`, `COMMIT`, and `ROLLBACK` commands.
+There are various restrictions on these transactions, primarily because the Storage and Compute layers require you to durably write before you can read from them.
+This makes read-after-write transactions challenging.
+Write-after-read transactions are implemented by **exclusively** performing all reads at the current timestamp and then writing at the next timestamp.
+Other read-only or write-only transactions may co-occur with a write-after-read transaction, but only one write-after-read transaction may span each transition from one time to the next.
+
+It is possible that multiple write-after-read transactions on disjoint collections could be made concurrent, but the technical requirements are open at the moment.
+
+Read-after-write, and general transactions are technically possible using advanced timestamps that allow for Adapter-private further resolution.
+All MZ collections support compensating update actions, and one can tentatively deploy updates and eventually potentially retract them, as long as others are unable to observe violations of atomicity.
+Further detail available upon request.
+
+It is critical that Adapter not deploy commands to Storage and Compute until a transaction has committed.
```

Comment:
I don't understand this. read-only and write-after-read txns both deploy commands before the transaction has committed.

Reply:
Ah, maybe what I mean to say is "not deploy `UpdateAndDowngrade` commands"? The DDL commands seem mostly harmless, I suppose?

Reply:
Yes that seems correct.